dns: properly trim response when EDNS is used

dnephin opened the pull request:
Fixes #9945

Best viewed by individual commit.

This PR refactors parts of the `DNSServer` to allow us to set the response code and trim the response in a single place. This fixes the problem of the DNS response exceeding the max size.

I notice we never called trim in `handlePtr` or `handleRecurse`. I wonder if we should be.

TODO:
* [x] changelog
* [x] add a new test specifically for this case
* [x] figure out if the last remaining TODO needs to be resolved, which might require returning the error down through the calls

dnephin commented on agent/dns.go:
```diff
@@ -838,23 +863,33 @@ func (d *DNSServer) trimDomain(query string) string {
-	}
-	if structs.IsErrNoDCPath(err) || structs.IsErrQueryNotFound(err) {
+	case errors.Is(err, errNoAnswer):
+		// TODO: why do we return success if the answer is empty?
+		return dns.RcodeSuccess
```
I don't think we need to resolve this TODO, but I added it because I don't fully understand this case. The existing comments say:

> If the answer is empty and the response isn't truncated, return not found

However, we were never setting the response code to `RcodeNameError`. Maybe the existing comment is misleading? The one thing we did for this case was set the SOA response, and we still do that, so this preserves existing behaviour.

Does anyone know why this we set SOA for this case, but not the response code?

dnephin commented on agent/dns_test.go:
```diff
@@ -545,6 +545,7 @@ func TestDNS_NodeLookup_CNAME(t *testing.T) {
 
 	m := new(dns.Msg)
 	m.SetQuestion("google.node.consul.", dns.TypeANY)
+	m.SetEdns0(8192, true)
```
This test started to fail because we are now trimming every response instead of only some of them. So I had to `setEdns0` to increase the size of the response.

I guess this is probably expected behaviour, and is now more correct? I'm honestly not sure.

dhiaayachi replied:
The spec state that without eDNS we should trim to 512 by default and this is what we do now, So I think we are good

dnephin commented on agent/dns.go:
```diff
@@ -1896,6 +1900,7 @@ func (d *DNSServer) resolveCNAME(cfg *dnsConfig, name string, maxRecursionLevel
 
 		req.SetQuestion(name, dns.TypeANY)
+		// TODO: handle error response
 		d.dispatch("udp", nil, req, resp, maxRecursionLevel-1)
```
This is the last TODO to resolve. Looking at it again, I guess we only ever return the `resp.Answer` here, so ignoring the error might be ok, since the error only impacts the `MsgHdr` and we won't send that anyway.

dnephin commented on agent/dns.go:
```diff
@@ -838,23 +859,33 @@ func (d *DNSServer) trimDomain(query string) string {
 	qType := req.Question[0].Qtype
 	if qType != dns.TypeANY && qType != dns.TypeA && qType != dns.TypeAAAA && qType != dns.TypeTXT {
-		return
+		return errQueryRefused
```
This is also new. I guess previously we returned some partially incomplete response? Maybe it had a success Rcode? I'm not really sure, but I'm pretty sure it was wrong.  Now we should be returning a response with `RcodeRefused`, which seems better.

dhiaayachi replied:
not sure, in that case it mean we have the wrong query type (if I understand right). Like I explained in the other comment we should return an empty success if the Q type is wrong but the node exist.

dnephin replied:
Yup, you are definitely right. I will fix this as well.

dhiaayachi commented on agent/dns.go:
```diff
@@ -838,23 +859,33 @@ func (d *DNSServer) trimDomain(query string) string {
-	}
-	if structs.IsErrNoDCPath(err) || structs.IsErrQueryNotFound(err) {
+	case errors.Is(err, errNoAnswer):
+		// TODO: why do we return success if the answer is empty?
```
This is to avoid caching of error response for other query types. if we return an error a recursor will cache it and use it even for other query types

dnephin replied:
Makes sense now that I've learned a bunch about DNS this week! I'll remove this TODO.

dhiaayachi reviewed (commented):
@dnephin I added the missing test and added an error check for the CNAME.
The only part I'm not sure about is the change in the error handling change in `rCodeFromError`.
dns error are a bit tricky because of how response caching is done in recursors and also the fact that a cached response could be returned to a different type of request (different Qtype for example)

dnephin commented on agent/dns_test.go:
```diff
@@ -6658,6 +6658,47 @@ func TestDNS_PreparedQuery_AgentSource(t *testing.T) {
+		m.SetEdns0(2048, true)
+
+		c := new(dns.Client)
+		r, _, err := c.Exchange(m, a.DNSAddr())
+		require.NoError(t, err)
+		require.Len(t, r.Answer, 53)
```
I think if we call `r.Pack()` we can also test that the returned byte buffer is less than 2048, which might be good to verify as well in case of the size of the records changes somehow.

From my reading of `WriteMsg` the only other thing that can get added is `Server.TsigSecret`, which it looks like we don't set right now, so `Pack` should be accurate.

dhiaayachi replied:
yes I like the idea, I will add this

dnephin commented:
I'm going to rebase this PR to get the verify-ci check working, just an FYI because it will require a force push.

dnephin commented on agent/dns.go:
```diff
@@ -1358,13 +1361,11 @@ func (d *DNSServer) preparedQueryLookup(cfg *dnsConfig, network, datacenter, que
-	if len(resp.Answer) == 0 && !resp.Truncated {
-		d.addSOA(cfg, resp)
-		return
+	if len(resp.Answer) == 0 {
+		return errNoAnswer
 	}
+	return nil
```
Looking at this part of the code changes, I'm wondering if the handling of SOA is still correct in this branch. There is a similar branch above in `serviceLookup`.

Previously we would only set SOA if we had no answers, where as now we'll set it in both cases (when the prepared query or service lookup returns no instances, and when it returns some service instances).

@blake do you know in what cases would we not want to set the `SOA` record on the response.NS section? Only for recurse, or are there other cases where setting the SOA record would be incorrect?

dhiaayachi replied:
I changed the code a bit to avoid trimming the answer after adding SOA, we only add SOA in certain conditions + no trimming happened (otherwise we will bust the response size limit).

dnephin replied:
Ok, I looked into this again and I think it was mostly correct. The only problem I found was that my fix in the commit with title "dns: correct rcode for qtype not supported" I returned `errNoAnswer` (which sets SOA) instead of `nil` (which doesn't). They both use the same RCode, but only one will set SOA.

I have no idea if this is correct, but at least it matches the old behaviour. I've amended the commit which introduced the problem to properly return a `nil` error when the Rcode is not supported by `nodeLookup`.

blake replied:
For unsupported qtypes, I believe the better response is to send `errNoAnswer` (or `errNoData` as I suggested renaming to).

https://umbrella.cisco.com/blog/nxdomain-nodata-debugging-dns-dual-stacked-hosts offers a good explanation as to why a NODATA response should be returned if resource records exist for other query types. Section 2.2 from RFC 2308 (linked in another comment) explains why its recommended to add the SOA.

> @blake do you know in what cases would we not want to set the SOA record on the response.NS section? Only for recurse, or are there other cases where setting the SOA record would be incorrect?

It took me a bit of staring at the code before I realized that response.Ns actually refers  to the the authority section of the DNS response. [[1]] :-)

After reading through a few RFCs, the SOA should definitely be omitted in negative responses for non-authoritative resources (i.e., recursive queries). It is also recommended that the SOA be added to the authority section [[2]] of a negative response (NODATA or NXDOMAIN) for resources which the the server is authoritative [[3]].

> I have no idea if this is correct,…

Based on my understanding of the linked RFCs, this seems to be the correct behavior.

[1]: https://pkg.go.dev/github.com/miekg/dns#Msg
[2]: https://datatracker.ietf.org/doc/html/rfc2181#section-7.1
[3]: https://datatracker.ietf.org/doc/html/rfc1034#section-4.3.4

dnephin commented on agent/dns.go:
```diff
@@ -1899,9 +1899,13 @@ func (d *DNSServer) resolveCNAME(cfg *dnsConfig, name string, maxRecursionLevel
+			d.addSOA(cfg, resp)
+		}
+		resp.SetRcode(req, rCode)
 		return resp.Answer
```
@dhiaayachi  I think  the challenge with this TODO is that we only return `resp.Answer`.

`SetRcode and `addSOA` change different parts of the `resp`, they don't modify the `Answer` section, so when we return `Answer` here we throw away the `Rcode` and `NS` sections that we modified.

My TODO comment was not very complete, but I think the question we need to answer is: Do we need to care about the  error returned from `dispatch` ?

If we need to care, then `resolveCNAME` needs to return either the full `resp` or some other struct to include all of that information.

dnephin replied:
Another possibility is that `dispatch` is just the wrong abstraction here, but if that's the case it'll be more work to clean up I think.

dhiaayachi replied:
yes you are right, I also when through the call flow and we don't really care about dispatch failing. it end up returning a nil answer which mean that at the end we won't add any extra fields.

dnephin commented:
I rebased this PR with the latest main branch, and removed commit bb42aea23cab822f37c33d7384efe3d52e6aa9a3 (which was causing test failures), and merged the commit+revert to cancel those out.

blake commented on agent/dns.go:
```diff
@@ -849,23 +869,30 @@ func (d *DNSServer) trimDomain(query string) string {
-// computeRCode Return the DNS Error code from Consul Error
-func (d *DNSServer) computeRCode(err error) int {
-	if err == nil {
+// rCodeFromError return the DNS Error code an error
```
```suggestion
// rCodeFromError return the appropriate DNS response code for a given error
```

dnephin commented on agent/dns.go:
```diff
@@ -610,9 +607,37 @@ func (d *DNSServer) parseDatacenter(labels []string, datacenter *string) bool {
+var errECSNotGlobal = fmt.Errorf("ECS response is not global")
+var errNameNotFound = fmt.Errorf("DNS name not found")
+
+// errNoAnswer is used to indicate that the response should set SOA, and the
+// success response code.
+var errNoAnswer = fmt.Errorf("no DNS Answer")
```
This is much better, thank you! Definitely agree we should make the domain terminology as much as possible.

dnephin commented on agent/dns.go:
```diff
@@ -1189,14 +1189,24 @@ func trimUDPResponse(req, resp *dns.Msg, udpAnswerLimit int) (trimmed bool) {
+		if cfg.EnableTruncate {
+			resp.Truncated = true
+		}
+		d.logger.Debug("DNS response too large, truncated",
+			"protocol", network,
+			"question", req.Question,
+			"records", fmt.Sprintf("%d/%d", len(resp.Answer), originalNumRecords),
+			"size", fmt.Sprintf("%d/%d", resp.Len(), originalSize),
+		)
```
I made a small change to this log message:
1. to always emit it when trimmer, even if we aren't setting the truncated field. I believe technically we truncate in both cases, but we only indicate it in the response when the setting is enabled (seems like strange behaviour to me)
2. removed TCP from the log message and added a protocol field to indicate tcp or udp

dnephin reviewed (commented):
@dhiaayachi your changes LGTM! Thank you for all of your help on this

dhiaayachi reviewed (approved):
LGTM!! Thanks @dnephin !!

hc-github-team-consul-core commented:
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/412032.

hc-github-team-consul-core commented:
:cherries::white_check_mark: Cherry pick of commit 1c8ac9cd4bda5b265fad4d7776f9592fa72721a0 onto `release/1.10.x` succeeded!

hc-github-team-consul-core commented:
:cherries::x: Cherry pick of commit 1c8ac9cd4bda5b265fad4d7776f9592fa72721a0 onto `release/1.9.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/412032)

hc-github-team-consul-core commented:
:cherries::x: Cherry pick of commit 1c8ac9cd4bda5b265fad4d7776f9592fa72721a0 onto `release/1.8.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/412032)
